File: modules/expo-bluesky-gif-view/ios/ExpoBlueskyGifViewModule.swift
import ExpoModulesCore
import SDWebImage
import SDWebImageWebPCoder

public class ExpoBlueskyGifViewModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBlueskyGifView")

    OnCreate {
      SDImageCodersManager.shared.addCoder(SDImageGIFCoder.shared)
    }

    AsyncFunction("prefetchAsync") { (sources: [URL]) in
      SDWebImagePrefetcher.shared.prefetchURLs(sources, context: Util.createContext(), progress: nil)
    }

    View(GifView.self) {
      Events(
        "onPlayerStateChange"
      )

      Prop("source") { (view: GifView, prop: String) in
        view.source = prop
      }

      Prop("placeholderSource") { (view: GifView, prop: String) in
        view.placeholderSource = prop
      }

      Prop("autoplay") { (view: GifView, prop: Bool) in
        view.autoplay = prop
      }

      AsyncFunction("toggleAsync") { (view: GifView) in
        view.toggle()
      }

      AsyncFunction("playAsync") { (view: GifView) in
        view.play()
      }

      AsyncFunction("pauseAsync") { (view: GifView) in
        view.pause()
      }
    }
  }
}

File: modules/expo-bluesky-gif-view/android/src/main/java/expo/modules/blueskygifview/ExpoBlueskyGifViewModule.kt
package expo.modules.blueskygifview

import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoBlueskyGifViewModule : Module() {
  override fun definition() =
    ModuleDefinition {
      Name("ExpoBlueskyGifView")

      AsyncFunction("prefetchAsync") { sources: List<String> ->
        val activity = appContext.currentActivity ?: return@AsyncFunction
        val glide = Glide.with(activity)

        sources.forEach { source ->
          glide
            .download(source)
            .diskCacheStrategy(DiskCacheStrategy.DATA)
            .submit()
        }
      }

      View(GifView::class) {
        Events(
          "onPlayerStateChange",
        )

        Prop("source") { view: GifView, source: String ->
          view.source = source
        }

        Prop("placeholderSource") { view: GifView, source: String ->
          view.placeholderSource = source
        }

        Prop("autoplay") { view: GifView, autoplay: Boolean ->
          view.autoplay = autoplay
        }

        AsyncFunction("playAsync") { view: GifView ->
          view.play()
        }

        AsyncFunction("pauseAsync") { view: GifView ->
          view.pause()
        }

        AsyncFunction("toggleAsync") { view: GifView ->
          view.toggle()
        }
      }
    }
}
